apps/server/src/routers/api/v1/upload.rs

```diff
@@ -0,0 +1,474 @@
+/// fails then an error is returned.
+fn validate_zip_file(zip_file: &mut ZipFile) -> APIResult<()> {
+	/// Any file extension not in this list will trigger an error
+	const ALLOWED_EXTENSIONS: &[&str] = &[
+		"cbr", "cbz", "epub", "pdf", "xml", "json", "png", "jpg", "jpeg", "webp", "gif",
+		"heif", "jxl", "avif",
+	];
+
+	/// Any inferred mime type not in this list will trigger an error
+	const ALLOWED_TYPES: &[&str] = &[
+		"application/zip",
+		"application/vnd.comicbook+zip",
+		"application/vnd.comicbook-rar",
+		"application/epub+zip",
+		"application/pdf",
+		"application/xml",
+		"application/json",
+		"image/png",
+		"image/jpeg",
+		"image/webp",
+		"image/gif",
+		"image/heif",
+		"image/jxl",
+		"image/avif",
+	];
+
+	let Some(enclosed_path) = zip_file.enclosed_name() else {
+		return Err(APIError::InternalServerError(
+			"Series zip contained a malformed path".to_string(),
+		));
+	};
+
+	let extension = enclosed_path
+		.extension()
+		.and_then(|ext| ext.to_str())
+		.map(str::to_ascii_lowercase)
+		.ok_or_else(|| {
+			APIError::InternalServerError(format!(
+				"Expected zip contents {enclosed_path:?} to have an extension."
+			))
+		})?;
+
+	if !ALLOWED_EXTENSIONS.contains(&extension.as_str()) {
+		return Err(APIError::InternalServerError(format!(
+			"Zip contents {enclosed_path:?} has a disallowed extension, permitted extensions are: {ALLOWED_EXTENSIONS:?}"
+		)));
+	}
+
+	// Read first five bytes from which to infer content type
+	let mut magic_bytes = [0u8; 5];
+	zip_file.read_exact(&mut magic_bytes).map_err(|_| {
+		APIError::InternalServerError(
+			"Failed to read first five bytes of zip file.".to_string(),
+		)
+	})?;
+
+	// TODO - Evaluate this. How often might valid uploads still fail due to infer?
+	let inferred_type = infer::get(&magic_bytes)
+		.ok_or(APIError::InternalServerError(format!(
+			"Unable to infer type for zip contents {enclosed_path:?}"
+		)))?
+		.mime_type();
+
+	if !ALLOWED_TYPES.contains(&inferred_type) {
+		return Err(APIError::InternalServerError(format!(
+			"Zip contents {enclosed_path:?} has a disallowed mime type: {inferred_type}, permitted types are: {ALLOWED_TYPES:?}"
+		)));
+	}
+
+	Ok(())
+}
+
+/// Returns `true` if a parameter specifying a path from another path contains no parent directory components.
+///
+/// Upload paths for books are recieved as a path offset, where the actual path is constructed as
+/// `{library_path}/{offset}`. This could be a security vulnerability if someone sent an upload with
+/// a path containing a `..` to push the path back to the parent directoy. This could be used to escape
+/// the library and upload things elsewhere. It also means that accepting only paths that start with the
+/// library path isn't sufficient.
+///
+/// This function will reject any paths that include a parent directory component. There is unlikely to be
+/// any circumstance where a client sending one would be appropriate anyhow.
+fn is_subpath_secure(params: &str) -> bool {
```

Comment:
I don't think so, unless I misunderstand your meaning. The `place_at` path will sometimes include more than one path component (e.g., if the user is uploading to a nested directory).

Reply:
So basically we'd reject `/`?

My thinking is still that we might not want that. First, the `place_at` variable is always going to be either appended to the library path or, alternatively, contain it. In either of those cases the final path can't be root.

I also think that `/` is a valid value for the variable - it would be equivalent to saying "put it at the library root."

Reply:
> So basically we'd reject /?

Yeah, basically - but it was mostly an open question

> [/] would be equivalent to saying "put it at the library root

Ah, good point! Thank you! I think it is moot then, I forgot that detail.